### PR TITLE
[IMP] http: Sessions implicit deactivation

### PR DESCRIPTION
**Pull request by tbe-odoo**
- Store a token inside sessions to allow implicit session deactivation when needed.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/res/res_users.py [hunk 1]
```diff
@@ -510,6 +514,36 @@ def check(cls, db, uid, passwd):
...
 
+    def _get_session_token_fields(self):
+        token_fields = set(USER_PRIVATE_FIELDS + ['active', 'password_crypt'])
+        return set(self.fields_get_keys()).intersection(token_fields)
+
+    @tools.ormcache('uid', 'sid')
+    def _compute_session_token(self, sid, uid):
+        """ Compute a session token given a session id and a user id """
+        # retrieve user's password crypt
+        session_fields = ', '.join(self._get_session_token_fields()) # SQL-i prone
```
Comment (KangOl) [inline odoo/addons/base/res/res_users.py]: It should be safe to sql-i.
But you can quote identifier using [quote_ident](http://initd.org/psycopg/docs/extensions.html#psycopg2.extensions.quote_ident) or using [psycopg2.sql](http://initd.org/psycopg/docs/sql.html)

File: odoo/addons/base/res/res_users.py [hunk 2]
```diff
@@ -510,6 +514,36 @@ def check(cls, db, uid, passwd):
...
+        self.env.cr.execute("""SELECT %s, (SELECT value FROM ir_config_parameter WHERE key='database.secret')
+                                FROM res_users
+                                WHERE id=%%s""" % (session_fields), (uid,))
+        if self.env.cr.rowcount != 1:
+            self._invalidate_session_cache()
+            return False
+        # 'stringify' the fields to use as the key
+        data_fields = str(sorted(self.env.cr.fetchone()))
```
Comment (KangOl) [inline odoo/addons/base/res/res_users.py]: Why sort result?
That should be `session_fields` that should be sorted.

File: odoo/addons/base/res/res_users.py [hunk 3]
```diff
@@ -510,6 +514,36 @@ def check(cls, db, uid, passwd):
...
+        # generate hmac key
+        key = (u'%s-%s' % (uid, data_fields)).encode('utf-8')
+        # hmac the session id
+        data = sid.encode()
```
Comment (KangOl) [inline odoo/addons/base/res/res_users.py]: force encoding to `utf-8` (sid is taken as argument, being non-ascii)

File: odoo/service/security.py [hunk 4]
```diff
@@ -11,3 +11,16 @@ def login(db, login, password):
...
+
+def compute_session_token(db, sid, uid):
+    with odoo.registry(db).cursor() as cr:
+        self = odoo.api.Environment(cr, uid, {})['res.users']
+        return self._compute_session_token(sid, uid)
+
+def check_session(db, sid, uid, token):
+    with odoo.registry(db).cursor() as cr:
+        self = odoo.api.Environment(cr, uid, {})['res.users']
+        if self._compute_session_token(sid, uid) == token:
```
Comment (KangOl) [inline odoo/service/security.py]: use [odoo.tools.misc.consteq()](https://github.com/odoo/odoo/blob/93c4c5cfb1945be6afdb3367b9f5b73f7a2c4f89/odoo/tools/misc.py#L1193-L1198)

File: odoo/addons/base/res/res_users.py [hunk 5]
```diff
@@ -510,6 +514,36 @@ def check(cls, db, uid, passwd):
         finally:
             cr.close()
 
+    def _get_session_token_fields(self):
+        token_fields = set(USER_PRIVATE_FIELDS + ['active', 'password_crypt'])
```
[the remark this answers is not in the document]
Comment (xmo-odoo) [reply odoo/addons/base/res/res_users.py]: Also a possibly cleaner expression would be `set(USER_PRIVATE_FIELDS) | {'active', 'password_crypt'})`

File: odoo/addons/base/res/res_users.py [hunk 6]
```diff
@@ -510,6 +514,36 @@ def check(cls, db, uid, passwd):
...
+        if self.env.cr.rowcount != 1:
+            self._invalidate_session_cache()
+            return False
+        # 'stringify' the fields to use as the key
+        data_fields = str(self.env.cr.fetchone())
+        # generate hmac key
+        key = (u'%s-%s' % (self.id, data_fields)).encode('utf-8')
+        # hmac the session id
+        data = sid.encode('utf-8')
+        h = hmac.new(key, data, sha1)
```
Comment (xmo-odoo) [inline odoo/addons/base/res/res_users.py]: Not technically necessary but given task 31589 (phase out sha1) maybe use a sha2 hash instead?

File: odoo/http.py [hunk 7]
```diff
@@ -1055,7 +1055,9 @@ def check_security(self):
         """
         if not self.db or not self.uid:
             raise SessionExpiredException("Session expired")
-        security.check(self.db, self.uid, self.password)
+        # here we check if the session is still valid
+        if not security.check_session(self.db, self.sid, self.uid, self.session_token):
```
Comment (xmo-odoo) [inline odoo/http.py]: Maybe just pass the session object and check_session can check whatever it wants? Same for compute_session_token?

**Comment by yelizariev** [comment]
Can ``database.secret`` be read via ORM, please?

```
+        self.env.cr.execute("""SELECT %s, (SELECT value FROM ir_config_parameter WHERE key='database.secret')
+                                FROM res_users
+                                WHERE id=%%s""" % (session_fields), (self.id,))

```

It broke multi-company module (it's in the top of app store), which make value field computable (via ir.property)

**Comment by odony** [comment]
@yelizariev Are you talking about [this code](https://github.com/it-projects-llc/misc-addons/blob/11.0/ir_config_parameter_multi_company/models/ir_config_parameter.py)?

It looks this module is breaking one of the basic rules of Odoo module encapsulation: it turns a stored field into a non-stored field. This is only working by chance, as long as no other module needs direct access to the columns for _any reason_.

The patch here is low-level and very sensitive, and one of the goal was to avoid being impacted by side-effects of ORM magic and third-party modules.
For instance, some of the fields in `_get_session_token_fields` are sensitive and protected from normal ORM read access. More such fields will appear in the future.

So, while technically we could make a special case for accessing the `database.secret`, and take it out of the SQL query, I'd rather avoid it. It will be a source of side-effects and future bugs. Especially if the only reason for this exception is a third-party module that breaks an important design principle of Odoo modules.
I know how it sounds, and I'm sorry for that, but I'm sure you realize that if you did the same thing  in a random core Odoo model (say, a model from the `account` module), you'd be bound to break things.

Short of re-designing your module in a way that does not break modularity principles, a workaround is to at least ensure that the original stored columns of ir.config.parameter remain available for direct SQL access, even if you don't use them. This patch should work as long as you keep the `key` and `value` columns.

**Comment by yelizariev** [comment]
Thanks for the reply. I understand your point. I'll try to find some solution. ``self.env['ir.config_parameter'].sudo().get_param('database.secret')``, would be the simplest way, though.

**Comment by iledarn** [comment]
@odony By the way, could you please share all the ``basic rules of Odoo module encapsulation`` with us? Where can I get to know with them? Thanks.

**Comment by odony** [comment]
@iledarn I'm not sure these are spelled out anywhere, as it's mostly common sense and good practices of object-oriented design. In OO programming, inheritance patterns tend to conflict with [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)).
As a result, components that inherit from other components should make sure they do not disrupt the normal behavior and API of these base components, as this would compromise the robustness of the system.

Applied to Odoo modules, it means that modules that inherit components for other modules should try to preserve the fundamental features and API of the inherited components (models, methods, fields, views, reports, etc.)

This includes, but is not limited to:
- not removing models
- not removing database columns from inherited models (unstoring a stored column)
- not adding constraints to inherited models that cannot be satisfied without the inheritance (e.g. new required columns)
- not changing the signature / return type of inherited methods

One rule of thumb to figure this out is to remember that the framework loads modules as layers, based on their dependencies:

> _A module that does not break encapsulation is a module whose layer can be removed from the registry without breaking the underlying layers. Such a module can also be easily uninstalled without requiring explicit reconstruction operations of the components it was inheriting._


And this is not just a theoretical point of view. During registry initialization, and maintenance operations such as installation/upgrade/removal of modules, the registry is _really_ loaded in multiple passes, starting with the modules with no dependencies. At each intermediary step, the framework needs to be able to use the CRUD methods and the API of the already-loaded modules, in order to perform the maintenance operations.
This won't be possible if modules that aren't yet loaded have significantly broken the encapsulation of the lower modules (e.g. missing tables, columns, etc.) It *may* work by chance, e.g. because the bad module happens to be in the same pass as its deps, but that's a very brittle situation that could easily break.

_Note: technically, the framework takes some precautions during the intermediary passes, such as ignoring the inherited views installed by not-yet-loaded-modules - and of course their Python logic is not loaded yet._

**Pull request merged**